### PR TITLE
Added examples of the GoF patterns and implemented one

### PR DESCRIPTION
**Pull request by Dimdim28**

### CODE REVIEW

File: oop.md [hunk 1]
```diff
@@ -0,0 +1,179 @@
+# Singleton
```
Comment (pie3phobic) [inline oop.md:1]: Yes, that's a good example of a Singleton design patter. It ensures that there is only one instance of a class and provides a global point of access to that instance. In this case, the `SingletonInstance` class creates and returns a single instance of the `AxiosInstance` class, which can be accessed globally through the `getInstance` method. The private constructor ensures that no new instances of `SingletonInstance` can be created from outside the class. GJ!

File: oop.md [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+      state.Error = action.error.message;
+    });
+  },
+});
+
+export const { setCurrentPage } = coursesSlice.actions;
+
+export default coursesSlice.reducer;
+```
+# Observer
```
Comment (pie3phobic) [inline oop.md:120]: Yes, I agree, in Redux, the store can be seen as the observable object, and the components that use useSelector and useDispatch can be seen as the observers. The store notifies the components of state changes, and the components notify the store of user actions.

**Review by pie3phobic — APPROVED** [review]
Analysed your code, the `couses-page.tsx` seems seems to be a good example of code that follows solid principles: 
1. Single Responsibility Principle - this component has a single responsibility of rendering a list of courses and pagination. It is not responsible for fetching data, handling errors, or updating the store. Instead, it delegates these responsibilities to other functions and hooks.

2. Open/Closed Principle - The component is open for extension, but closed for modification. For example, if we want to add a filter to the list of courses, we can create a new component that filters the courses and passes the filtered courses to this component. This way, we can add new features without modifying the existing code.
3. Dependency Inversion Principle - instead of directly importing specific modules, the code imports hooks from a custom appHooks module which abstracts the details of how the hooks are implemented, allowing the CoursesPage component to depend on abstractions rather than concrete implementations. This makes the code more modular and easier to test, as changes to the implementation of the hooks will not affect the CoursesPage component as long as the abstraction remains the same.
Nice work!

**Review by dima-cherednyk — APPROVED** [review]
GJ =)

**Review by HryhoriiUstymenko — APPROVED** [review]
Good job. Thank you for detailed pattern explanation and code examples!

**Pull request merged**